src/archivematica/storage_service/common/backends.py

```diff
@@ -95,17 +115,53 @@ def decode_token(token: str) -> Any:
         return info
 
     def create_user(self, user_info: dict[str, Any]) -> User:
+        """Create a new user when authentication was successful."""
+        role = self.get_user_role(user_info)
+        if role is None:
+            return None
+
         user = super().create_user(user_info)
         for attr, value in user_info.items():
             setattr(user, attr, value)
-        self.set_user_role(user)
+        user.set_role(role)
         return user
 
     def update_user(self, user: User, user_info: dict[str, Any]) -> User:
-        self.set_user_role(user)
+        """Updates the user's role only if the setting allows roles to be set from OIDC claims."""
+        if self.OIDC_OP_SET_ROLES_FROM_CLAIMS:
+            role = self.get_user_role(user_info)
+            if role is None:
+                return None
+            user.set_role(role)
         return user
 
-    def set_user_role(self, user: User) -> None:
-        # TODO: use user claims accessible via user's authentication tokens.
-        role = roles.promoted_role(roles.USER_ROLE_READER)
-        user.set_role(role)
+    def get_user_role(self, user_info: dict[str, Any]) -> Optional[str]:
+        """
+        Returns a promoted role if one of the OIDC token claims matches a valid role.
+        Returns None if no valid role is found.
+        """
+        if self.OIDC_OP_SET_ROLES_FROM_CLAIMS:
+            claim_path = self.OIDC_OP_ROLE_CLAIM_PATH.split(".")
+            role = user_info
+
+            # Traverse nested claim path to get role(s)
+            for key in claim_path:
+                if isinstance(role, dict):
+                    role = role.get(key)
+                else:
+                    role = None
+                    break
+
+            # If role is a list, find the first valid one
+            if isinstance(role, list):
+                for r in role:
+                    if r in self.VALID_ROLES:
+                        return roles.promoted_role(r)
```

Comment:
Do providers like Keycloak guarantee order of the realm roles of the user? If I understand this correctly it'll return the first role it matches not necessarily the one with more permissions. Should that fact be stated in the README?

Reply:
This is a good point and I was wondering how best to handle this. I have altered `get_user_role()` so that it returns the highest-permission valid role found in the OIDC token claims based on the order of the roles in `USER_ROLES` in `roles.py`

Reply:
Understood and I appreciate the new test!

Should we add a comment above `roles.USER_ROLES` to state the list should maintain a descending level of permissions from now on?

Reply:
Added comment in recent commit.